File: .claude/skills/startup-multi-device-analysis/scripts/hypothesis_tests.py
#!/usr/bin/env python3
"""Cross-pass hypothesis tests for one device's campaign.

Consumes the per-iteration JSON dumps produced by variance_analysis.py --json for each
campaign pass, plus the campaign log (battery temps), and prints the evidence for/against:

  H1  within-pass outliers = little-core placement (needs the device's real cluster map)
  H2  pass-level fast/slow alternation, uniform CPU inflation, block-resume sections ~2x
  H3  persisted-config-load bimodal: iter000 (fresh install, no cached config) fast
  H4  off-window fluctuators: power-service-registration binder stalls; native-lib IO

Run this per device BEFORE any cross-device or cross-arm comparison: if H2 fires, only
matching-state passes may be compared.

Usage: python3 hypothesis_tests.py <campaign-dir>   (expects pass1.json..passN.json + campaign.log)

Set LITTLE_CPUS (comma-separated cpu ids, default "0,1,2,3") to the little cluster's cpu
ids from device_probe.py's topology JSON. The default is wrong on any device whose little
cluster is not cpu0-3 and it fails silently, so set it explicitly per device.
"""

import json
import os
import re
import statistics
import sys

SLOW_DELTA_MS = 4.0        # floor; effective threshold is max(this, 10% of pass median)
                           # so "slow" scales across device tiers, whose median windows can
                           # differ by an order of magnitude between flagship and entry
CL0_MAJORITY = 0.5
CL0_ABSENT = 0.10          # H1 falsifier: slow iteration with <10% cluster-0 residency
CFGLOAD_FAST_MS = 5.0      # H3: below this = "no cached config" fast path
POWER_STALL_MS = 10.0
LITTLE_CPUS = {int(x) for x in os.environ.get("LITTLE_CPUS", "0,1,2,3").split(",")}

PURE_CPU = ["emb-span-service-init", "emb-otel-tracer-init"]
BLOCK_RESUME = ["emb-config-service-init", "emb-payload-source-init",
                "emb-essential-service-init", "emb-post-init", "emb-delivery-init"]


def main() -> int:
    camp = sys.argv[1]
    passes = []
    for i in range(1, 20):
        p = os.path.join(camp, f"pass{i}.json")
        if not os.path.exists(p):
            break
        with open(p) as f:
            passes.append(json.load(f))
    if not passes:
        print("no passN.json files found", file=sys.stderr)
        return 1
    temps = parse_temps(os.path.join(camp, "campaign.log"))

    print(f"hypothesis tests over {len(passes)} passes, "
          f"{sum(len(p) for p in passes)} iterations total")
    print()

    # ---- H2: pass-level behavior --------------------------------------------------
    print("H2. pass-level windows (ms) and battery temps")
    print(f"  {'pass':>4}{'n':>5}{'p50':>8}{'p90':>8}{'max':>8}{'run p50':>9}"
          f"{'temp start':>12}{'temp end':>10}")
    meds = []
    for i, data in enumerate(passes):
        wins = [d["window_ms"] for d in data]
        runs = [sum(d["cpu_ms"].values()) for d in data]
        meds.append(statistics.median(wins))
        t0, t1 = temps.get(i + 1, (None, None))
        print(f"  {i + 1:>4}{len(wins):>5}{statistics.median(wins):>8.1f}"
              f"{pctl(wins, 90):>8.1f}{max(wins):>8.1f}{statistics.median(runs):>9.1f}"
              f"{fmt(t0):>12}{fmt(t1):>10}")
    print(f"  pass medians sequence: {' -> '.join(f'{m:.1f}' for m in meds)}")
    fastest = min(range(len(passes)), key=lambda i: meds[i])
    slowest = max(range(len(passes)), key=lambda i: meds[i])
    print(f"  fastest pass {fastest + 1} ({meds[fastest]:.1f}), "
          f"slowest pass {slowest + 1} ({meds[slowest]:.1f}), "
          f"swing {meds[slowest] - meds[fastest]:.1f} ms")
    print("  slow/fast section-median ratios (H2 predicts pure-CPU ~1x, block-resume ~2x):")
    for c in PURE_CPU + BLOCK_RESUME:
        f_ = med_section(passes[fastest], c)
        s_ = med_section(passes[slowest], c)
        kind = "pure-CPU" if c in PURE_CPU else "block-resume"
        print(f"    {c:<38}{kind:<14}{f_:>7.2f}{s_:>7.2f}  ratio {s_ / f_:>5.2f}x")
    print()

    # ---- H1: cluster placement ----------------------------------------------------
    print("H1. little-core placement vs window")
    all_delta, all_share = [], []
    for i, data in enumerate(passes):
        wins = [d["window_ms"] for d in data]
        med = statistics.median(wins)
        shares = [cl0_share(d) for d in data]
        r = pearson(wins, shares)
        c0 = [w for w, s in zip(wins, shares) if s > CL0_MAJORITY]
        c1 = [w for w, s in zip(wins, shares) if s <= CL0_MAJORITY]
        print(f"  pass {i + 1}: r={r:>5.2f}  little-majority n={len(c0):>3} "
              f"mean={mean_or(c0):>6}  cluster1-majority n={len(c1):>3} mean={mean_or(c1):>6}")
        all_delta.extend(w - med for w in wins)
        all_share.extend(shares)
    print(f"  pooled r(window-delta-vs-pass-median, little-share) = "
          f"{pearson(all_delta, all_share):.2f}   n={len(all_delta)}")
    slow_iters = [(i + 1, j, d) for i, data in enumerate(passes)
                  for j, d in enumerate(data)
                  if d["window_ms"] - statistics.median([x["window_ms"] for x in data])
                  > max(SLOW_DELTA_MS,
                        0.10 * statistics.median([x["window_ms"] for x in data]))]
    on_c0 = [t for t in slow_iters if cl0_share(t[2]) > CL0_MAJORITY]
    falsifiers = [t for t in slow_iters if cl0_share(t[2]) < CL0_ABSENT]
    print(f"  slow iterations (delta > +{SLOW_DELTA_MS:.0f} ms): {len(slow_iters)} total, "
          f"{len(on_c0)} little-majority, {len(falsifiers)} FALSIFIERS "
          f"(<{CL0_ABSENT:.0%} little-core)")
    for pn, j, d in falsifiers[:12]:
        print(f"    pass{pn} iter{j:03d}: window {d['window_ms']:.1f}, "
              f"little {cl0_share(d):.0%}, cfgload {d['dur'].get('emb-persisted-config-load', 0):.1f}, "
              f"spansvc {d['dur'].get('emb-span-service-init', 0):.1f}")
    print()

    # ---- H3: persisted-config-load bimodality --------------------------------------
    print("H3. persisted-config-load: iter000 (fresh install) vs the rest, ms")
    print(f"  {'pass':>4}{'iter000':>9}{'rest min':>10}{'rest p50':>10}{'rest max':>10}"
          f"{'rest <5ms':>11}")
    for i, data in enumerate(passes):
        c0 = data[0]["dur"].get("emb-persisted-config-load")
        rest = [d["dur"].get("emb-persisted-config-load") for d in data[1:]]
        rest = [v for v in rest if v is not None]
        fast_rest = len([v for v in rest if v < CFGLOAD_FAST_MS])
        print(f"  {i + 1:>4}{c0:>9.2f}{min(rest):>10.2f}{statistics.median(rest):>10.2f}"
              f"{max(rest):>10.2f}{fast_rest:>11}")
    print()

    # ---- H4: off-window fluctuators -------------------------------------------------
    print("H4. off-window fluctuators")
    print(f"  {'pass':>4}{'power p50':>10}{'power max':>10}{'stalls>10':>10}"
          f"{'natlib io p50':>15}{'sighand io p50':>15}{'snap max':>10}")
    for i, data in enumerate(passes):
        power = [d["dur"].get("emb-power-service-registration") for d in data]
        power = [v for v in power if v is not None]
        stalls = len([v for v in power if v > POWER_STALL_MS])
        nat_io = [io_ms(d, "emb-load-embrace-native-lib") for d in data]
        sig_io = [io_ms(d, "emb-install-native-crash-signal-handlers") for d in data]
        snap = [d["dur"].get("emb-snapshot-session", 0) for d in data]
        print(f"  {i + 1:>4}{statistics.median(power):>10.2f}{max(power):>10.2f}{stalls:>10}"
              f"{statistics.median(nat_io):>15.2f}{statistics.median(sig_io):>15.2f}"
              f"{max(snap):>10.2f}")
    return 0


def cl0_share(d):
    tot = sum(d["cpu_ms"].values()) or 1.0
    return sum(v for k, v in d["cpu_ms"].items() if int(k) in LITTLE_CPUS) / tot


def io_ms(d, section):
    st = d["states"].get(section, {})
    return (st.get("D", 0) + st.get("D+io", 0) + st.get("DK", 0) + st.get("DK+io", 0)
            + st.get("S+io", 0))


def med_section(data, c):
    return statistics.median([d["dur"][c] for d in data if c in d["dur"]])


def pctl(vals, p):
    s = sorted(vals)
    return s[min(len(s) - 1, int(round(p / 100 * (len(s) - 1))))]


def mean_or(vals):
    if not vals:
        return "--"
    return f"{statistics.mean(vals):.1f}"


def fmt(t):
    if t is None:
        return "--"
    return f"{t:.1f}C"


def parse_temps(log_path):
    temps = {}
    if not os.path.exists(log_path):
        return temps
    with open(log_path) as f:
        for ln in f:
            m = re.search(r"pass (\d+)/\d+ starting, battery ([\d.]+)", ln)
            if m:
                temps.setdefault(int(m.group(1)), [None, None])[0] = float(m.group(2))
            m = re.search(r"pass (\d+) done in .* battery ([\d.]+)", ln)
            if m:
                temps.setdefault(int(m.group(1)), [None, None])[1] = float(m.group(2))
    return {k: tuple(v) for k, v in temps.items()}


def pearson(xs, ys):
    mx, my = statistics.mean(xs), statistics.mean(ys)
    num = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    den = (sum((x - mx) ** 2 for x in xs) * sum((y - my) ** 2 for y in ys)) ** 0.5
    if den == 0:
        return float("nan")
    return num / den


if __name__ == "__main__":
    sys.exit(main())

File: startup-tools/src/main/kotlin/io/embrace/startup/analysis/HypothesisTests.kt
package io.embrace.startup.analysis

import io.embrace.startup.analysis.VarianceAnalysis.Record
import io.embrace.startup.core.stats.Descriptive
import io.embrace.startup.core.stats.Quantile
import io.embrace.startup.core.text.PyFormat
import java.nio.file.Files
import java.nio.file.Path

/**
 * `hypothesis_tests.py`: cross-pass evidence for one device's campaign.
 *
 * - H2 pass-level fast/slow alternation, uniform CPU inflation, block-resume sections ~2×
 * - H1 within-pass outliers = little-core placement (needs the device's real cluster map)
 * - H3 persisted-config-load bimodal: iter000 (fresh install, no cached config) fast
 * - H4 off-window fluctuators: power-service-registration binder stalls; native-lib IO
 *
 * Run per device BEFORE any cross-device or cross-arm comparison: if H2 fires, only matching-state
 * passes may be compared. Note the label/threshold mismatch carried over from the Python: the
 * "slow iterations (delta > +4 ms)" line uses an effective threshold of max(4 ms, 10% of the pass
 * median), so "slow" scales across device tiers.
 */
object HypothesisTests {

    const val SLOW_DELTA_MS: Double = 4.0
    const val CL0_MAJORITY: Double = 0.5
    const val CL0_ABSENT: Double = 0.10
    const val CFGLOAD_FAST_MS: Double = 5.0
    const val POWER_STALL_MS: Double = 10.0

    val PURE_CPU: List<String> = listOf("emb-span-service-init", "emb-otel-tracer-init")
    val BLOCK_RESUME: List<String> = listOf(
        "emb-config-service-init",
        "emb-payload-source-init",
        "emb-essential-service-init",
        "emb-post-init",
        "emb-delivery-init",
    )

    /** Battery temperatures at pass start and end, keyed by pass number, from `campaign.log`. */
    data class Temps(val start: Double?, val end: Double?)

    fun parseTemps(log: Path): Map<Int, Temps> {
        if (!Files.exists(log)) return emptyMap()
        val starts = HashMap<Int, Double>()
        val ends = HashMap<Int, Double>()
        Files.readAllLines(log).forEach { ln ->
            START.find(ln)?.let { starts[it.groupValues[1].toInt()] = it.groupValues[2].toDouble() }
            DONE.find(ln)?.let { ends[it.groupValues[1].toInt()] = it.groupValues[2].toDouble() }
        }
        return (starts.keys + ends.keys).associateWith { Temps(starts[it], ends[it]) }
    }

    fun report(
        passes: List<List<Record>>,
        temps: Map<Int, Temps> = emptyMap(),
        littleCpus: Set<Int> = VarianceAnalysis.DEFAULT_LITTLE_CPUS,
    ): String {
        val out = StringBuilder()
        out.line("hypothesis tests over ${passes.size} passes, ${passes.sumOf { it.size }} iterations total")
        out.line()
        out.h2(passes, temps)
        out.h1(passes, littleCpus)
        out.h3(passes)
        out.h4(passes)
        return out.toString()
    }

    private fun StringBuilder.line(s: String = ""): StringBuilder = append(s).append('\n')

    private fun StringBuilder.h2(passes: List<List<Record>>, temps: Map<Int, Temps>) {
        line("H2. pass-level windows (ms) and battery temps")
        line(
            "  ${"pass".padStart(W4)}${"n".padStart(W5)}${"p50".padStart(W8)}${"p90".padStart(W8)}${"max".padStart(W8)}" +
                "${"run p50".padStart(W9)}${"temp start".padStart(W12)}${"temp end".padStart(W10)}",
        )
        val meds = ArrayList<Double>()
        passes.forEachIndexed { i, data ->
            val wins = data.map { it.window() }
            val runs = data.map { it.cpuMs.values.sum() }
            val med = Quantile.median(wins.sorted())
            meds.add(med)
            val t = temps[i + 1]
            line(
                "  ${(i + 1).toString().padStart(W4)}${wins.size.toString().padStart(W5)}${f(med, 1, W8)}" +
                    "${f(pctl(wins, P90), 1, W8)}${f(wins.max(), 1, W8)}${f(Quantile.median(runs.sorted()), 1, W9)}" +
                    "${fmtTemp(t?.start).padStart(W12)}${fmtTemp(t?.end).padStart(W10)}",
            )
        }
        line("  pass medians sequence: ${meds.joinToString(" -> ") { PyFormat.fixed(it, 1) }}")
        val fastest = meds.indices.minByOrNull { meds[it] } ?: 0
        val slowest = meds.indices.maxByOrNull { meds[it] } ?: 0
        line(
            "  fastest pass ${fastest + 1} (${PyFormat.fixed(meds[fastest], 1)}), " +
                "slowest pass ${slowest + 1} (${PyFormat.fixed(meds[slowest], 1)}), " +
                "swing ${PyFormat.fixed(meds[slowest] - meds[fastest], 1)} ms",
        )
        line("  slow/fast section-median ratios (H2 predicts pure-CPU ~1x, block-resume ~2x):")
        (PURE_CPU + BLOCK_RESUME).forEach { c ->
            val fast = medSection(passes[fastest], c)
            val slow = medSection(passes[slowest], c)
            val kind = if (c in PURE_CPU) "pure-CPU" else "block-resume"
            line("    ${c.padEnd(W38)}${kind.padEnd(W14)}${f(fast, 2, W7)}${f(slow, 2, W7)}  ratio ${f(slow / fast, 2, W5)}x")
        }
        line()
    }

    private fun StringBuilder.h1(passes: List<List<Record>>, littleCpus: Set<Int>) {
        line("H1. little-core placement vs window")
        val allDelta = ArrayList<Double>()
        val allShare = ArrayList<Double>()
        passes.forEachIndexed { i, data ->
            val wins = data.map { it.window() }
            val med = Quantile.median(wins.sorted())
            val shares = data.map { cl0Share(it, littleCpus) }
            val r = Descriptive.pearson(wins, shares)
            val c0 = wins.filterIndexed { j, _ -> shares[j] > CL0_MAJORITY }
            val c1 = wins.filterIndexed { j, _ -> shares[j] <= CL0_MAJORITY }
            line(
                "  pass ${i + 1}: r=${f(r, 2, W5)}  little-majority n=${c0.size.toString().padStart(W3)} " +
                    "mean=${meanOr(c0).padStart(W6)}  cluster1-majority n=${c1.size.toString().padStart(W3)} " +
                    "mean=${meanOr(c1).padStart(W6)}",
            )
            wins.forEach { allDelta.add(it - med) }
            allShare.addAll(shares)
        }
        line(
            "  pooled r(window-delta-vs-pass-median, little-share) = " +
                "${PyFormat.fixed(Descriptive.pearson(allDelta, allShare), 2)}   n=${allDelta.size}",
        )
        val slowIters = passes.flatMapIndexed { i, data ->
            val med = Quantile.median(data.map { it.window() }.sorted())
            val threshold = maxOf(SLOW_DELTA_MS, TENTH * med)
            data.mapIndexedNotNull { j, d -> if (d.window() - med > threshold) Triple(i + 1, j, d) else null }
        }
        val onC0 = slowIters.filter { cl0Share(it.third, littleCpus) > CL0_MAJORITY }
        val falsifiers = slowIters.filter { cl0Share(it.third, littleCpus) < CL0_ABSENT }
        line(
            "  slow iterations (delta > +${PyFormat.fixed(SLOW_DELTA_MS, 0)} ms): ${slowIters.size} total, " +
                "${onC0.size} little-majority, ${falsifiers.size} FALSIFIERS (<${PyFormat.percent0(CL0_ABSENT)} little-core)",
        )
        falsifiers.take(MAX_FALSIFIERS).forEach { (pn, j, d) ->
            line(
                "    pass$pn iter${j.toString().padStart(3, '0')}: window ${PyFormat.fixed(d.window(), 1)}, " +
                    "little ${PyFormat.percent0(cl0Share(d, littleCpus))}, " +
                    "cfgload ${PyFormat.fixed(d.dur["emb-persisted-config-load"] ?: 0.0, 1)}, " +
                    "spansvc ${PyFormat.fixed(d.dur["emb-span-service-init"] ?: 0.0, 1)}",
            )
        }
        line()
    }

    private fun StringBuilder.h3(passes: List<List<Record>>) {
        line("H3. persisted-config-load: iter000 (fresh install) vs the rest, ms")
        line(
            "  ${"pass".padStart(W4)}${"iter000".padStart(W9)}${"rest min".padStart(W10)}${"rest p50".padStart(W10)}" +
                "${"rest max".padStart(W10)}${"rest <5ms".padStart(W11)}",
        )
        passes.forEachIndexed { i, data ->
            val c0 = data[0].dur["emb-persisted-config-load"] ?: Double.NaN
            val rest = data.drop(1).mapNotNull { it.dur["emb-persisted-config-load"] }
            val fastRest = rest.count { it < CFGLOAD_FAST_MS }
            val sorted = rest.sorted()
            line(
                "  ${(i + 1).toString().padStart(W4)}${f(c0, 2, W9)}${f(sorted.first(), 2, W10)}" +
                    "${f(Quantile.median(sorted), 2, W10)}${f(sorted.last(), 2, W10)}${fastRest.toString().padStart(W11)}",
            )
        }
        line()
    }

    private fun StringBuilder.h4(passes: List<List<Record>>) {
        line("H4. off-window fluctuators")
        line(
            "  ${"pass".padStart(W4)}${"power p50".padStart(W10)}${"power max".padStart(W10)}${"stalls>10".padStart(W10)}" +
                "${"natlib io p50".padStart(W15)}${"sighand io p50".padStart(W15)}${"snap max".padStart(W10)}",
        )
        passes.forEachIndexed { i, data ->
            val power = data.mapNotNull { it.dur["emb-power-service-registration"] }.sorted()
            val stalls = power.count { it > POWER_STALL_MS }
            val natIo = data.map { ioMs(it, "emb-load-embrace-native-lib") }.sorted()
            val sigIo = data.map { ioMs(it, "emb-install-native-crash-signal-handlers") }.sorted()
            val snap = data.map { it.dur["emb-snapshot-session"] ?: 0.0 }
            line(
                "  ${(i + 1).toString().padStart(W4)}${f(Quantile.median(power), 2, W10)}${f(power.last(), 2, W10)}" +
                    "${stalls.toString().padStart(W10)}${f(Quantile.median(natIo), 2, W15)}" +
                    "${f(Quantile.median(sigIo), 2, W15)}${f(snap.max(), 2, W10)}",
            )
        }
    }

    fun cl0Share(d: Record, littleCpus: Set<Int>): Double {
        val tot = d.cpuMs.values.sum().takeIf { it != 0.0 } ?: 1.0
        return d.cpuMs.filterKeys { it.toInt() in littleCpus }.values.sum() / tot
    }

    fun ioMs(d: Record, section: String): Double {
        val st = d.states[section] ?: emptyMap()
        return (st["D"] ?: 0.0) + (st["D+io"] ?: 0.0) + (st["DK"] ?: 0.0) + (st["DK+io"] ?: 0.0) + (st["S+io"] ?: 0.0)
    }

    /** `pctl`: nearest-rank on `round(p/100 * (n-1))`, Python's half-even `round`. Kept for this report only. */
    fun pctl(vals: List<Double>, p: Int): Double {
        val s = vals.sorted()
        val idx = Math.rint(p / PERCENT * (s.size - 1)).toInt()
        return s[minOf(s.size - 1, idx)]
    }

    private fun medSection(data: List<Record>, c: String): Double =
        Quantile.median(data.mapNotNull { it.dur[c] }.sorted())

    private fun meanOr(vals: List<Double>): String =
        if (vals.isEmpty()) "--" else PyFormat.fixed(PyFormat.exactMean(vals), 1)

    private fun fmtTemp(t: Double?): String = if (t == null) "--" else "${PyFormat.fixed(t, 1)}C"

    private fun f(x: Double, decimals: Int, width: Int) = PyFormat.fixed(x, decimals).padStart(width)

    private fun Record.window(): Double = requireNotNull(windowMs) { "$trace has no window_ms" }

    private val START = Regex("pass (\\d+)/\\d+ starting, battery ([\\d.]+)")
    private val DONE = Regex("pass (\\d+) done in .* battery ([\\d.]+)")
    private const val P90 = 90
    private const val TENTH = 0.10
    private const val MAX_FALSIFIERS = 12
    private const val PERCENT = 100.0
    private const val W3 = 3
    private const val W4 = 4
    private const val W5 = 5
    private const val W6 = 6
    private const val W7 = 7
    private const val W8 = 8
    private const val W9 = 9
    private const val W10 = 10
    private const val W11 = 11
    private const val W12 = 12
    private const val W14 = 14
    private const val W15 = 15
    private const val W38 = 38
}
